src/Xamarin.Android.Build.Tasks/Utilities/XAAssemblyResolver.cs

```diff
@@ -0,0 +1,333 @@
+	/// specific assemblies.  Each ABI has its own set of directories to search.
+	/// <summary>
+	public IDictionary<AndroidTargetArch, ICollection<string>> AbiSearchDirectories { get; } = new Dictionary<AndroidTargetArch, ICollection<string>> ();
+
+	readonly List<MemoryMappedViewStream> viewStreams = new List<MemoryMappedViewStream> ();
+	bool disposed;
+	TaskLoggingHelper log;
+	bool loadDebugSymbols;
+	ReaderParameters readerParameters;
+	readonly Dictionary<string, CacheEntry> cache;
+
+	public XAAssemblyResolver (TaskLoggingHelper log, bool loadDebugSymbols, ReaderParameters? loadReaderParameters = null)
+	{
+		this.log = log;
+		this.loadDebugSymbols = loadDebugSymbols;
+		this.readerParameters = loadReaderParameters ?? new ReaderParameters();
+
+		cache = new Dictionary<string, CacheEntry> (StringComparer.OrdinalIgnoreCase);
+	}
+
+	public AssemblyDefinition Resolve (string fullName, ReaderParameters? parameters = null)
+	{
+		return Resolve (AssemblyNameReference.Parse (fullName), parameters);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name)
+	{
+		return Resolve (name, null);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		return Resolve (AndroidTargetArch.None, name, parameters);
+	}
+
+	public AssemblyDefinition Resolve (AndroidTargetArch arch, AssemblyNameReference name, ReaderParameters? parameters = null)
+	{
+		string shortName = name.Name;
+		if (cache.TryGetValue (shortName, out CacheEntry? entry)) {
+			return SelectAssembly (arch, name.FullName, entry, loading: false);
+		}
+
+		if (arch == AndroidTargetArch.None) {
+			return FindAndLoadFromDirectories (arch, FrameworkSearchDirectories, name, parameters);
+		}
+
+		if (!AbiSearchDirectories.TryGetValue (arch, out ICollection<string>? directories) || directories == null) {
+			throw CreateLoadException (name);
+		}
+
+		return FindAndLoadFromDirectories (arch, directories, name, parameters);
+	}
+
+	AssemblyDefinition FindAndLoadFromDirectories (AndroidTargetArch arch, ICollection<string> directories, AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		string? assemblyFile;
+		AssemblyDefinition? candidate = null;
+		foreach (string dir in directories) {
+			if ((assemblyFile = SearchDirectory (name.Name, dir)) != null) {
+				AssemblyDefinition? loaded = Load (arch, assemblyFile);
+				if (Array.Equals (loaded?.Name.MetadataToken, name.MetadataToken)) {
+					return loaded;
+				}
+				candidate = candidate ?? loaded;
+			}
+		}
+
+		if (candidate != null)
+			return candidate;
+
+		throw CreateLoadException (name);
+	}
+
+	static FileNotFoundException CreateLoadException (AssemblyNameReference name)
+	{
+		return new FileNotFoundException ($"Could not load assembly '{name}'.");
+	}
+
+	static string? SearchDirectory (string name, string directory)
+	{
+		if (Path.IsPathRooted (name) && File.Exists (name)) {
```

Comment:
A lot of the code is copied from the JI `DirectoryAssemblyResolver`, I kept portions of code unmodified, assuming they were there for a reason and to keep the two resolvers as close as possible in terms of behavior.

Reply:
Also, it doesn't hurt to handle this case. Today, we get relative paths but it's external input, passed to the task by MSBuild, so it can very well change to absolute paths one day.